smart/solvers.py

```diff
@@ -493,29 +550,52 @@ def init_petsc_vector(self, j, assemble=False):
             V.assemble()
         return V
 
-    def Jijk_name(self, i, j, k=None):
-        ij = i * self.dim + j
+    def Jijk_name(self, i: int, j: int, k: Optional[int] = None):
+        """
+        Get a string representation of an entry of the Jacobian.
+
+        Args:
+            i: Row index
+            j: Column index
+            k: If the Jacobian entry is a sum of forms, get the name
+                of the domain in the `k`th entry.
+        """
         if k is None:
             return (
                 f"J{i}{j} = dF[{self.active_compartment_names[i]}]"
                 f"/du[{self.active_compartment_names[j]}]"
             )
         else:
+            ij = i * self.dim + j
+            if (num_entries := len(self.Jforms_all[ij])) <= k:
```

Comment:
Nice usage of the walrus operator `:=`. Note that is a python3.8 feature, which mean that we are implicitly dropping support for python3.7.

Reply:
Maybe we should add this constraint to the `pyproject.toml`, i.e https://packaging.python.org/en/latest/specifications/declaring-project-metadata/#requires-python ?

Reply:
> Nice usage of the walrus operator `:=`. Note that is a python3.8 feature, which means that we are implicitly dropping support for python3.7.

3.7 is eol June/July, so i think that would be ok

Reply:
Agree👍